pkg/agent/controller/noderoute/node_route_controller.go

```diff
@@ -405,6 +415,21 @@ func (c *Controller) addNodeRoute(nodeName string, node *corev1.Node) error {
 		// Does not help to return an error and trigger controller retries.
 		return nil
 	}
+
+	nodesHaveSamePodCIDR, _ := c.installedNodes.ByIndex(nodeRouteInfoPodCIDRIndexName, node.Spec.PodCIDR)
+	// PodCIDRs can be released from deleted Nodes and allocated to new Nodes. For server side, it won't happen that a
+	// PodCIDR is allocated to more than one Node at any point. However, for client side, if a resync happens to occur
+	// when there are Node creation and deletion events, the informer will generate the events in a way that all
+	// creation events come before deletion ones even they actually happen in the opposite order on the server side.
+	// See https://github.com/kubernetes/kubernetes/blob/v1.18.2/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L503-L512
+	// Therefore, a PodCIDR may appear in a new Node before the Node that previously owns it is removed. To ensure the
+	// stale routes, flows, and relevant cache of this podCIDR are removed appropriately, we wait for the Node deletion
+	// event to be processed before proceeding, or the route installation and uninstallation operations may override or
+	// conflict with each other.
+	if len(nodesHaveSamePodCIDR) > 0 {
+		return fmt.Errorf("podCIDR %s for Node %s is duplicate with Node %s", node.Spec.PodCIDR, nodeName, nodesHaveSamePodCIDR[0].(*nodeRouteInfo).nodeName)
```

Comment:
Should we mention that this error could be temporary? or something like addNoteRoute is deferred

Reply:
Not a bad idea.

Reply:
thanks for the suggestion, done